testgen/src/tester.rs

```diff
@@ -1,77 +1,196 @@
+        path.to_str().and_then(|path| TestEnv::new(path))
+    }
+
+    pub fn current_dir(&self) -> &str {
+        &self.current_dir
+    }
+
+    pub fn logln(&self, msg: &str) -> Option<()> {
+        println!("{}", msg);
+        self.full_path("_log").and_then(|full_path| {
+            fs::OpenOptions::new()
+                .create(true)
+                .append(true)
+                .open(full_path)
+                .ok()
+                .and_then(|mut file| file.write_all((String::from(msg) + "\n").as_bytes()).ok())
+        })
+    }
+
+    pub fn logln_to(&self, msg: &str, rel_path: &str) -> Option<()> {
+        println!("{}", msg);
+        self.full_path(rel_path).and_then(|full_path| {
+            fs::OpenOptions::new()
+                .create(true)
+                .append(true)
+                .open(full_path)
+                .ok()
+                .and_then(|mut file| file.write_all((String::from(msg) + "\n").as_bytes()).ok())
+        })
+    }
+
+    /// Read a file from a path relative to the environment current dir into a string
+    pub fn read_file(&self, rel_path: &str) -> Option<String> {
+        self.full_path(rel_path)
+            .and_then(|full_path| fs::read_to_string(&full_path).ok())
+    }
+
+    /// Write a file to a path relative to the environment current dir
+    pub fn write_file(&self, rel_path: &str, contents: &str) -> Option<()> {
+        self.full_path(rel_path)
+            .and_then(|full_path| fs::write(full_path, contents).ok())
+    }
+
+    /// Parse a file from a path relative to the environment current dir as the given type
+    pub fn parse_file_as<T: DeserializeOwned>(&self, rel_path: &str) -> Option<T> {
+        self.read_file(rel_path)
+            .and_then(|input| serde_json::from_str(&input).ok())
+    }
+
+    /// Copy a file from the path outside environment into the environment current dir
+    /// Returns the relative path of the file, or None if copying was not successful
+    pub fn copy_file_from(&self, path: &str) -> Option<String> {
+        let path = Path::new(path);
+        if !path.is_file() {
+            return None;
         }
+        path.file_name().and_then(|name| {
+            name.to_str().and_then(|name| {
+                self.full_path(name)
+                    .and_then(|dest| fs::copy(path, dest).ok().map(|_| name.to_string()))
+            })
+        })
+    }
+
+    /// Copy a file from the path relative to the other environment into the environment current dir
+    /// Returns the relative path of the file, or None if copying was not successful
+    pub fn copy_file_from_env(&self, other: &TestEnv, path: &str) -> Option<String> {
+        other
+            .full_path(path)
+            .and_then(|full_path| self.copy_file_from(&full_path))
     }
 
     /// Convert a relative path to the full path from the test root
     /// Return None if the full path can't be formed
     pub fn full_path(&self, rel_path: &str) -> Option<String> {
```

Comment:
Why not return a `PathBuf` instead of an `Option<String>`?

Reply:
The problem with passing and returning `Path` / `PathBuf` is that this shifts the complexity from the library to its user. E.g., in another PR I have the following code:

```
    env.copy_file_from_env(root_env, "Lightclient_A_1.tla");
    env.copy_file_from_env(root_env, "Blockchain_A_1.tla");
    env.copy_file_from_env(root_env, "LightTests.tla");
    env.copy_file_from_env(root_env, &test.model);
```

With `&str` this is easy; If the second parameter is `Path`, it becomes a nightmare...

Reply:
Right, that makes sense.

Perhaps we can go with `path: impl AsRef<Path>` then? This allows both string-like types (&str, String) and path-like types (Path, PathBuf) and is thus a bit more flexible. It is also a common pattern in the standard library: https://users.rust-lang.org/t/when-to-use-asref-t-vs-t/9312/3

Reply:
That's indeed a good idea:) I have quickly tested, it seems to work. Thanks!

As a side-note: I find the current Rust APIs around Path/PathBuf/OsStr a bit over-complicated. Can't make a clear mental picture of it...

Reply:
Do you mean the API exposed by the types or the need for these types themselves?

If the latter: the gist of it is:

- `PathBuf` is an owned path (similar to what `String` is to strings, or `Vec<u8>` to byte arrays)
- `Path` is a borrowed reference to a path (similar to what `&str` is to strings, or `&[u8]` is to byte arrays)
- `OsString` is an owned platform-native path string (~ `String`)
- `OsStr` is a borrowed reference to a platform-native string (~ `&str`)

For the latter two, I feel like the doc of [`OsString`](https://doc.rust-lang.org/stable/std/ffi/struct.OsString.html) gives a good explanation for why the type is needed. [This post](https://fasterthanli.me/articles/working-with-strings-in-rust) is another good resource worth reading.

Reply:
Thank you for the explanation and the references, very useful. It's just the feeling that there are too many atoms in this string universe. The number of atoms, multiplied by the number of their interconnections is even bigger; hard to navigate;)  And also that feeling that this number could be reduced. Probably it will get easier as I learn more of the Rust standard libs, may be not. At least I am not alone:
* [pathbuf-and-path-why-not-string](https://users.rust-lang.org/t/pathbuf-and-path-why-not-string/28777)
* [is_working_with_paths_always_this_painful](https://www.reddit.com/r/rust/comments/7mu7q1/is_working_with_paths_always_this_painful/)

E.g., taking into account [PathBuf::into_os_str()](https://doc.rust-lang.org/std/path/struct.PathBuf.html#method.into_os_string) would it be a good idea just to merge the types `PathBuf`/`OsString`, and `Path`/`OsStr`? I don't know.